static/js/publisher/metrics/graphs/activeDevicesGraph.js

```diff
@@ -0,0 +1,736 @@
+        .append("div")
+        .attr("class", "p-tooltip");
+    }
+
+    if (this.options.area) {
+      renderArea();
+    } else {
+      renderLines();
+    }
+
+    renderAnnotations();
+
+    renderXAxis();
+    renderYAxis();
+
+    this.hasRendered = true;
+
+    return this;
+  }
+
+  tooltipTemplate(dateData, currentHoverKey) {
+    const tooltipRows = (dateData, currentHoverKey) => {
+      let dataArr = [];
+      let other = {
+        key: "other",
+        value: 0,
+        count: 0
+      };
+
+      let keys = Object.keys(dateData).filter(key => key !== "date");
+
+      if (this.options.graphType && this.options.graphType === "channel") {
+        keys = sortChannels(keys, {
+          defaultTrack: this.options.metricsDefaultTrack
+            ? this.options.metricsDefaultTrack
+            : "latest"
+        }).list;
+      }
+
+      keys.forEach(key => {
+        dataArr.push({
+          key: key,
+          value: dateData[key]
+        });
+      });
+
+      if (
+        (this.options.graphType && this.options.graphType !== "channel") ||
+        this.options.stacked === false
+      ) {
+        dataArr.sort((a, b) => b.value - a.value);
+      }
+
+      if (dataArr.length > 15) {
+        dataArr.forEach((item, index) => {
+          if (index >= 15) {
+            other.value += item.value;
+            other.count += 1;
+            other.key = `${other.count} other`;
+            item.skip = true;
+          }
+        });
+      }
+
+      // If we've added anything to the 'other' series,
+      // add it to the array
+      if (other.value > 0) {
+        dataArr.push(other);
+      }
+
+      return dataArr
+        .map(item => {
+          if (!item.skip) {
+            return [
+              `<span class="snapcraft-graph-tooltip__series${
+                item.key === currentHoverKey ? " is-hovered" : ""
+              }" title="${item.key}">`,
+              `<span class="snapcraft-graph-tooltip__series-name">${
+                item.key
+              }</span>`,
+              `<span class="snapcraft-graph-tooltip__series-color"${
+                !item.count
+                  ? `style="background:${this.colorScale(item.key)};"`
+                  : ""
+              }></span>`,
+              `<span class="snapcraft-graph-tooltip__series-value">${this.commaValue(
+                item.value
+              )}</span>`,
+              `</span>`
+            ].join("");
+          }
+        })
+        .join("");
+    };
+
+    return [
+      `<div class="p-tooltip p-tooltip--top-center" style="display: block; pointer-events: none;">`,
+      `<span class="p-tooltip__message" role="tooltip" style="display: block;">`,
+      `<span class="snapcraft-graph-tooltip__title">${this.tooltipTimeFormat(
+        dateData.date
+      )}</span>`,
+      tooltipRows(dateData, currentHoverKey),
+      `</span>`,
+      `</div>`
+    ].join("");
+  }
+
+  mouseMove(d, index, nodes) {
```

Comment:
It's clear that some (if not most) of the methods are meant for private use only (`mouseMove`, `tooltipTemplate`, etc). It would be nice to have those more clearly separated from public class API.

It's really hard to tell right now which methods are meant for public API use. `enableToopltip` seems to be public, but is `cancelTooltip` public or private?

This class contains such a lot of very specific code (that depens a lot on knowing d3 and our metrics data structure) that it's really hard to understand what it does internally, but also how to use it.
Any sort of documentation (even if just in comments above public API methods) would be helpful. And maybe prefixing private methods with `_` to make them distinct and less likely to be called from outside?